[IMP] web: discard unchanged record in editable list

mcm-odoo opened the pull request:
Before this commit, if we created a new row in an editable list view,
do not "touched" it and clicked out of the list then the record tried
to be saved and threw error notifications.
Now, that flow will discard the record to make it consistent with the
form view which discards the record when we leave the form.

task-2431691

aab-odoo commented on addons/account/static/tests/section_and_note_tests.js:
```diff
@@ -89,6 +89,8 @@ QUnit.module('section_and_note', {
         $tr1 = form.$('tr.o_data_row:eq(1)');
         await testUtils.dom.click($tr1.find('td.o_data_cell'));
+        await testUtils.nextTick(); // wait for discard
+        $tr1 = form.$('tr.o_data_row:eq(1)');
```
I still don't understand. There's no row to discard here, we just pass the first row in edition and then the second one

mcm-odoo replied:
clicking on the second row unselects the first with discardIfNotDirty option set to true

aab-odoo commented on addons/web/static/src/js/views/list/list_editable_renderer.js:
```diff
@@ -1826,7 +1845,11 @@ ListRenderer.include({
         }
 
-        this.unselectRow();
+        this.unselectRow({
```
i thought this would be the only place where the option should be set.

mcm-odoo replied:
I thought too but if we leave the view or change the page or clicking on a row while there is a new unchanged row then it throws an error

msh-odoo commented:
@mcm-odoo I already worked on it here: https://github.com/odoo/odoo/pull/64836 for task 2431691

mcm-odoo commented:
Hello @msh-odoo I know you already work on it, I took your commits, squashed them and cleaned a bit.
You're still the author of the new commit, I didn't change a lot of your code.
Also I tried to touch less modules than you did.

aab-odoo commented on addons/web/static/src/js/views/list/list_editable_renderer.js:
```diff
@@ -468,11 +468,14 @@ ListRenderer.include({
      *
+     * @param {Object} [options]
+     * @param {boolean} [options.discardIfNotDirty=false] set to true when
+     *   user clicks anywhere on the screen except save button.
```
this docstring says when/how the option is used. This is not what a docstring should be: it should rather explain what the option does.

aab-odoo commented on addons/web/static/src/js/views/list/list_editable_renderer.js:
```diff
@@ -487,12 +490,28 @@ ListRenderer.include({
-            });
+        return new Promise(async (resolve, reject) => {
+            const record = this._getRecord(recordID);
+            if (options.discardIfNotDirty && record.isNew()) {
+                await this.commitChanges(recordID);
+            }
+            if (options.discardIfNotDirty && record.isNew() && !record.isDirty()) {
+                this.trigger_up('mutexify', {
+                    action: () => {
+                        return this.trigger_up('discard_changes', {
+                            recordID: recordID,
+                            onSuccess: resolve,
+                            onFailure: reject,
+                        });
+                    },
+                });
+            } else {
+                this.trigger_up('save_line', {
+                    recordID: recordID,
+                    onSuccess: resolve,
+                    onFailure: reject,
+                });
+            }
```
```suggestion
            if (options.discardIfNotDirty && record.isNew()) {
                await this.commitChanges(recordID);
                if (!record.isDirty()) {
                    this.trigger_up('mutexify', {
                        action: () => {
                            return this.trigger_up('discard_changes', {
                                recordID: recordID,
                                onSuccess: resolve,
                                onFailure: reject,
                            });
                        },
                    });
                    return;
                }
            }
            this.trigger_up('save_line', {
                recordID: recordID,
                onSuccess: resolve,
                onFailure: reject,
            });
```

msh-odoo commented:
> Hello @msh-odoo I know you already work on it, I took your commits, squashed them and cleaned a bit.
> You're still the author of the new commit, I didn't change a lot of your code.
> Also I tried to touch less modules than you did.

@mcm-odoo thanks but I checked that this is the same task and you took my commits on which I worked but I confused due to different task ID give in PR, so I thought whether it is duplicate work, BTW thanks and sorry for misunderstanding :)

aab-odoo commented:
robodoo +

aab-odoo commented:
robodoo r+
